source/s3_client.c

```diff
@@ -1501,24 +1499,44 @@ static int s_s3_client_start_resolving_addresses_synced(struct aws_s3_client *cl
                                                 .shutdown_callback = s_s3_client_host_listener_shutdown_callback,
                                                 .user_data = client};
 
+    bool listener_already_exists = false;
+    bool error_occurred = false;
+
+    s_s3_client_lock_synced_data(client);
+
+    if (client->synced_data.host_listener != NULL) {
+        listener_already_exists = true;
+        goto unlock;
+    }
```

Comment:
Not a stupid question!  The issue is not a double add of the host listener.  It's the callback we feed into aws_host_resolver_resolve_host.  That callback needs to lock the mutex, which is usually okay.  But, if the host has already been resolved before we got here, the callback gets called synchronously, and then we have a double lock.  

But, because we do now have some of the function outside of the lock, we have to check to make sure another caller didn't add the listener in the time it took us to get here.  (Which is where this if comes in.)